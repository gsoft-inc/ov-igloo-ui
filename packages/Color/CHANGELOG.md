# @igloo-ui/color

## 0.1.1

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.1.0

### Minor Changes

- 2a6d8e5: Extracted the Color and Avatar components from the Tag component

  ## BREAKING CHANGE

  Restructuring HTML in the tag component as well as renaming onClose to OnRemove since it made more sense. onRemove also gains an id as a parameter.
