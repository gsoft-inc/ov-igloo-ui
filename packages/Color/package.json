{
  "name": "@igloo-ui/color",
  "version": "0.1.1",
  "main": "dist/Color.js",
  "module": "dist/Color.js",
  "types": "dist/Color.d.ts",
  "styles": "dist/color.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "classnames": "^2.3.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
