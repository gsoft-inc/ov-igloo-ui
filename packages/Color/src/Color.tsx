import * as React from "react";

import cx from "classnames";

import "./color.scss";

export type Size = "small" | "medium" | "large" | "xlarge";

export interface CustomCSSProps extends React.CSSProperties {
    "--ids-color-background"?: string;
    "--ids-color-text"?: string;
}

export interface ColorProps {
    /** The content of the color */
    children?: React.ReactNode;
    /** Add a specific class to the color */
    className?: string;
    /** The background color */
    color?: string;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** Whether or not the color is square */
    isSquare?: boolean;
    /** Specify the name that will be used to create initials */
    name?: string;
    /** Specifies the size of the color */
    size?: Size;
    /** The color of the text */
    textColor?: string;
}

const Color: React.FunctionComponent<ColorProps> = ({
    children,
    className,
    color,
    dataTest,
    isSquare = false,
    name,
    size = "small",
    textColor
}) => {
    const classes = cx(
        "ids-color",
        {
            [`ids-color--${size}`]: size, 
            "ids-color--square": isSquare
        },
        className
    );

    const initialsFromName = React.useMemo(() => {
        if (!name) {
            return null;
        }

        let initials = "";
        const cleanName = name.replace(/\s+/g, " ").trim();
        const splitName = cleanName.trim().split(" ");

        // If only one word, take the first 2 characters
        if (splitName.length === 1) {
            initials = splitName[0].slice(0, 2);
        }
        // If more than 1 word, take the first word's first character
        // then take the next first character of a segment who is alphadecimal
        else if (splitName.length > 1) {
            initials = splitName[0].slice(0, 1);
            for (let i = 1; i < splitName.length; i += 1) {
                const firstLetterOrSegment = splitName[i].slice(0, 1);
                if (firstLetterOrSegment.match(/\w/)) {
                    initials += firstLetterOrSegment;

                    break;
                }
            }

            if (initials.length < 2) {
                initials = splitName[0].slice(0, 2);
            }
        }

        return initials;
    }, [name]);

    const styleProps = React.useMemo(() => {
        const style: CustomCSSProps = {};
        if (color) {
            style["--ids-color-background"] = color;
        }

        if (textColor) {
            style["--ids-color-text"] = textColor;
        }

        return style;
    }, [color, textColor]);

    return (
        <div
            className={classes}
            data-test={dataTest}
            style={styleProps}
        >
            {children || initialsFromName}
        </div>
    );
};

export default Color;
