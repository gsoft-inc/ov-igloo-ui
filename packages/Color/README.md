# Color

A color can be used in place of an avatar. The background color and text color are customizable.

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/color` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/color
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/color
```

## Usage

Then to use the component in your code just import it!

```jsx
import Color from '@igloo-ui/color';

<Color
  color="#FFC3BA"
  textColor="#6B2837"
  size="large"
  name="Awesome Possum Team"
/>;
```
