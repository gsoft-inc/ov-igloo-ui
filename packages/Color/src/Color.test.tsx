/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';

import Color, { Size } from './Color';

const setup = (props = {}) => {
  return render(<Color dataTest="ids-color" {...props} />);
};

describe('Color', () => {
  const expectToBeOfSize = (component: RenderResult, size: Size) => {
    const color = component.container.firstChild;
    const expectedClass = 'ids-color--' + size;
    expect(color).toHaveClass(expectedClass);
  };

  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-color');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should render a small color', () => {
    const color = setup({ size: 'small' });

    expectToBeOfSize(color, 'small');
  });

  test('It should render a medium color', () => {
    const color = setup({ size: 'medium' });

    expectToBeOfSize(color, 'medium');
  });

  test('It should render a large color', () => {
    const color = setup({ size: 'large' });

    expectToBeOfSize(color, 'large');
  });

  test('It should render a xlarge color', () => {
    const color = setup({ size: 'xlarge' });

    expectToBeOfSize(color, 'xlarge');
  });
});
