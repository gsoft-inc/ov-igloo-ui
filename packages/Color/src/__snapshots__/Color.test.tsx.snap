// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-color ids-color--small ids-color--circular"
    data-test="ids-color"
  >
    <div
      class="ids-color__inner"
    />
  </div>
</DocumentFragment>
`;
